File: LAB7(5-2-19)/amicable.py
import threading
import time

x=0
y=0
s1=0
s2=0
def A(lock):
    lock.acquire()
    global x,s1
    print("thread name:",threading.current_thread().name,end="  ")
    x=int(input("Enter first number:"))

    for i in range (1,x):
        if(x%i ==0):
            s1+=i
    lock.release()



def B(lock):
    lock.acquire()
    global y,s2
    print("thread name:",threading.current_thread().name,end="  ")
    y=int(input("Enter second number:"))
    for i in range (1,y):
        if(y%i ==0):
            s2+=i
    lock.release()




lock=threading.Lock()

csa=threading.Thread(target=A, args=(lock,),name="CSA")

csb=threading.Thread(target=B, args=(lock,),name="CSB")

csa.start()
# time.sleep(2)
csb.start()

csa.join()
csb.join()
if(x==s2 and y==s1):
    print("numbers are amicable")
else:
    print("numbers are not amicable")

File: LAB7(5-2-19)/amicable.java
import java.util.*;
class generate extends Thread
{
	int num,sum;
	public generate(int a)
	{
		num=a;
		sum=0;
	}
	public synchronized void run()
	{
		System.out.println("In Thread"+getName());
		for(int i=1;i<num;i++)
		{
			if(num%i==0)
			{
				sum+=i;
			}
		}
		//this.notify();
	}
	public int getnum()
	{
		return num;
	}
	public int getsum()
	{
		return sum;
	}
}
class amicable
{
	public static void main(String []args)
	{
		System.out.println("ENTER FIRST NUMBER::");
		Scanner sc=new Scanner(System.in);
		int a=sc.nextInt();
		System.out.println("ENTER SECOND NUMBER::");
		int b=sc.nextInt();
		generate g1=new generate(a);
		generate g2=new generate(b);
		g1.setName("CSA");
		g2.setName("CSB");
		g1.start();
		g2.start();
		/*synchronized(g1)
		{
			try
			{
				g1.wait();
			}
			catch(InterruptedException ex)
			{
				ex.printStackTrace();
			}
		}
		synchronized(g2)
		{
			try
			{
				g2.wait();
			}
			catch(InterruptedException ex)
			{
				ex.printStackTrace();
			}
		}*/
		try
		{
			g1.join();
			g2.join();
		}
		catch(InterruptedException ex)
		{
			ex.printStackTrace();
		}
		System.out.println("In Thread "+Thread.currentThread().getName());
		if((g1.getnum()==g2.getsum())  && (g2.getnum()==g1.getsum()))
		{
			System.out.println(g1.getnum()+" and "+g2.getnum()+" are amicable.");
		}
		else
		{
			System.out.println(g1.getnum()+" and "+g2.getnum()+" are not amicable.");
		}
	}
}
